### PR TITLE
Feat: Download organization information - 186

### PR DESCRIPTION
**Pull request by hamed-valiollahi**
This PR introduces the ability for IDIR users to download organization information into a spreadsheet.

**Testing:**
- Added backend unit tests for the export route.
- Added frontend unit tests for download button state changes.
- Added Cypress E2E tests for the download functionality.

**Technical Enhancements:**
- SpreadsheetBuilder: Implemented a new `SpreadsheetBuilder` class capable of building spreadsheets in xlsx, xls, or csv formats. This class supports adding multiple sheets with custom styling.
- useApiService: Extended the `useApiService` to support file download capabilities, aligning with the new feature requirements.

### CODE REVIEW

File: backend/lcfs/web/api/organization/views.py [hunk 1]
```diff
@@ -28,6 +32,65 @@
...
+                organization.org_status.status.value,
+            ] for organization in organizations
+        ]
+
+        # Create a spreadsheet
+        builder = SpreadsheetBuilder(file_format = export_format)
+
+        builder.add_sheet(
+            sheet_name="Organizations",
+            columns=["ID", "Organization Name", "Compliance Units", "Registered"],
```
Comment (AlexZorkin) [inline backend/lcfs/web/api/organization/views.py:72]: Just add one more column here called 'In Reserve' so we have it when we get to compliance reporting.

File: backend/lcfs/utils/spreadsheet_builder.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        styles={'bold_headers': True, 'column_widths': [10, 20]}
+    )
+
+    # Second sheet
+    builder.add_sheet( ... )
+
+    file_content = builder.build_spreadsheet()
+    """
+    def __init__(self, file_format: str = "xls"):
+        self.file_format = file_format
```
Comment (AlexZorkin) [inline backend/lcfs/utils/spreadsheet_builder.py:30]: Could we add the file format validation to the __init__ method to catch this earlier?

File: backend/lcfs/utils/spreadsheet_builder.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+                sheet.col(col_index).width = 256 * width
+
+        for row_index, row in enumerate(sheet_data["rows"], start=1):
+            for col_index, value in enumerate(row):
+                # Apply left-aligned style for numbers, default style for others
+                cell_style = left_aligned_num_style if isinstance(value, (int, float)) else xlwt.XFStyle()
+                sheet.write(row_index, col_index, value, cell_style)
+
+    def _write_csv(self, output):
+        if self.sheets_data:
```
Comment (AlexZorkin) [inline backend/lcfs/utils/spreadsheet_builder.py:141]: Are you thinking we wouldn't have CSVs with more than one sheet_data? I'm fine with this but add a comment saying so.

File: backend/lcfs/utils/spreadsheet_builder.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        alignment = xlwt.Alignment()
+        alignment.horz = xlwt.Alignment.HORZ_LEFT
+        left_aligned_num_style.alignment = alignment
+
+        # Apply bold style for headers if required
+        header_style = bold_style if sheet_data["styles"].get("bold_headers") else default_style
+
+        for col_index, column in enumerate(sheet_data["columns"]):
+            sheet.write(0, col_index, column, header_style)
+            if "column_widths" in sheet_data["styles"] and len(sheet_data["styles"]["column_widths"]) > col_index:
```
Comment (AlexZorkin) [inline backend/lcfs/utils/spreadsheet_builder.py:130]: I'm thinking we try and make the column_widths dynamic based on the widest cell in each column. Could you try something like the following and let me know what you think.

```python
from openpyxl.utils import get_column_letter

def _auto_adjust_column_width_xlsx(self, writer, sheet_data):
    worksheet = writer.sheets[sheet_data["sheet_name"]]
    for column in worksheet.columns:
        max_length = 0
        column_letter = get_column_letter(column[0].column)

        for cell in column:
            try:
                if len(str(cell.value)) > max_length:
                    max_length = len(cell.value)
            except:
                pass

        adjusted_width = (max_length + 2)  # Adding 2 for a little extra space
        worksheet.column_dimensions[column_letter].width = adjusted_width

File: backend/lcfs/web/api/organization/views.py [hunk 5]
```diff
@@ -28,6 +32,65 @@
 get_async_db = dependencies.get_async_db_session
 
 
+@router.get("/export", response_class=StreamingResponse)
+@roles_required("Government")
+async def export(request: Request, db: AsyncSession = Depends(get_async_db)):
+    """
+    Endpoint to export information of all organization
```
Comment (AlexZorkin) [inline backend/lcfs/web/api/organization/views.py:39]: expand on this doc string to include what file format is and a brief on the spreadsheet builder

File: frontend/src/views/Organizations/Organizations.test.jsx [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    const downloadButton = screen.getByRole('button', { name: /Download Organization Information/i });
+    fireEvent.click(downloadButton);
+
+    // Wait for the download operation to complete and the component state to update
+    await waitFor(() => {
+        expect(downloadButton).toHaveTextContent(/Downloading Organization Information.../i);
+        expect(downloadButton).toBeDisabled();
+    });
+  });
+});
```
Comment (AlexZorkin) [inline frontend/src/views/Organizations/Organizations.test.jsx:50]: please add a download fail test, something like:

```js

it('shows an error message if the download fails', async () => {
  // Mock the download function to simulate a failure
  vi.mock('@/services/useApiService', () => ({
    useApiService: () => ({
      download: vi.fn().mockRejectedValueOnce(new Error('Download failed'))
    })
  }));

  renderComponent();

  const downloadButton = screen.getByRole('button', { name: /Download Organization Information/i });
  fireEvent.click(downloadButton);

  // Wait for the component to react to the error
  await waitFor(() => {
    const errorMessage = screen.getByText(/Download failed/i);
    expect(errorMessage).toBeInTheDocument();
  });
});

**Review by AlexZorkin — CHANGES_REQUESTED** [review]
Great spreadsheet class. This will be a good standard to use throughout the app. A few comments to address. ty

**Comment by hamed-valiollahi** [comment]
> Great spreadsheet class. This will be a good standard to use throughout the app. A few comments to address. ty

Thank you, Alex, for reviewing my code. I have made some improvements based on your feedback. Now, the code is significantly better.

File: backend/lcfs/web/api/organization/views.py (hunk 1 above)
Comment (hamed-valiollahi) [reply backend/lcfs/web/api/organization/views.py]: Sure, it's added.

File: backend/lcfs/utils/spreadsheet_builder.py (hunk 2 above)
Comment (hamed-valiollahi) [reply backend/lcfs/utils/spreadsheet_builder.py:37]: Of course! Thank you for pointing this out.

File: backend/lcfs/utils/spreadsheet_builder.py (hunk 3 above)
Comment (hamed-valiollahi) [reply backend/lcfs/utils/spreadsheet_builder.py:169]: I forgot to add a note for this! I've added a comment to mention it. Thank you, Alex!

File: backend/lcfs/utils/spreadsheet_builder.py (hunk 4 above)
Comment (hamed-valiollahi) [reply backend/lcfs/utils/spreadsheet_builder.py]: That's a great idea! I was thinking about it, but wasn't sure if I was allowed to do so. Now, it's better for both developers and users. Thanks for bringing it up.

File: backend/lcfs/web/api/organization/views.py (hunk 5 above)
Comment (hamed-valiollahi) [reply backend/lcfs/web/api/organization/views.py]: Of course!

File: frontend/src/views/Organizations/Organizations.test.jsx (hunk 6 above)
Comment (hamed-valiollahi) [reply frontend/src/views/Organizations/Organizations.test.jsx]: Sure, thanks for pointing this out. I've added it to both the end-to-end and front-end unit tests.

**Review by AlexZorkin — APPROVED** [review]

**Pull request merged**